src/Corax/Querying/Matches/SortingMatches/NumericalMaxHeapSorter.cs

```diff
@@ -0,0 +1,303 @@
+    {
+        _terms[0] = term;
+        _documents[0] = documentId;
+        MaxHeapify(0);
+    }
+
+    private void RemoveMax()
+    {
+        _terms[0] = _terms[_heapSize - 1];
+        _documents[0] = _documents[_heapSize - 1];
+        _heapSize--;
+
+        MaxHeapify(0);
+    }
+
+    public void Fill(Span<long> batchResults, ref ContextBoundNativeList<long> results, ref ContextBoundNativeList<float> scoreDestination, Span<float> scores)
+    {
+        ValidateMaxHeapStructure();
+        var start = results.Count;
+        results.EnsureCapacityFor(_heapSize);
+        int documentsToReturn = _heapSize;
+
+        var exposeScores = scoreDestination.HasContext && scores.Length != 0;
+        if (exposeScores)
+            scoreDestination.EnsureCapacityFor(_heapSize);
+        
+        while (documentsToReturn > 0)
+        {
+            results.AddUnsafe(batchResults[_documents[0]]);
+            
+            if (exposeScores)
+                scoreDestination.AddUnsafe(scores[_documents[0]]);
+            
+            RemoveMax();
+            documentsToReturn--;
+        }
+
+        Debug.Assert(_heapSize == 0, "_heapSize == 0");
+        // It's easier for us to operate on a MaxHeap, so we get the elements in descending order (locally).
+        // However, instead of rebuilding the heap as a MinHeap, let's add everything from the heap and call Reverse from the Span extension,
+        // which is a vectorized operation.
+        results.ToSpan().Slice(start).Reverse();
+    }
+
+    public void FillWithTerms(Span<long> batchResults, ref ContextBoundNativeList<long> results, ref ContextBoundNativeList<TTermType> terms, ref ContextBoundNativeList<float> scoreDestination, Span<float> scores)
+    {
+        ValidateMaxHeapStructure();
+
+        var total = _heapSize;
+        var startDocuments = results.Count;
+        var startTerms = terms.Count;
+
+        var exposeScores = scoreDestination.HasContext && scores.Length != 0;
+        if (exposeScores)
+            scoreDestination.EnsureCapacityFor(_heapSize);
+        
+        results.EnsureCapacityFor(_heapSize);
+        terms.EnsureCapacityFor(_heapSize);
+        int documentsToReturn = _heapSize;
+
+        while (documentsToReturn > 0)
+        {
+            results.AddUnsafe(batchResults[_documents[0]]);
+            terms.AddUnsafe(_terms[0]);
+            if (exposeScores)
+                scoreDestination.AddUnsafe(scores[_documents[0]]);
+            
+            RemoveMax();
+            documentsToReturn--;
+        }
+
+        Debug.Assert(_heapSize == 0, "_heapSize == 0");
+        // It's easier for us to operate on a MaxHeap, so we get the elements in descending order (locally).
+        // However, instead of rebuilding the heap as a MinHeap, let's add everything from the heap and call Reverse from the Span extension,
+        // which is a vectorized operation.
+        results.ToSpan().Slice(startDocuments, total).Reverse();
```

Comment:
Explain this for someone that is coming cold on what's going on here.

Reply:
done